### PR TITLE
8265244: Remove useless comparation in LibraryCallKit::inline_vector_convert()

### PR DESCRIPTION
**Pull request by Wanghuang-Huawei**
* aarch64 can only accept `VectorReinterpret` with 64/128 bits.
* I will fix this bug by adding a rule for `VectorReinterpret` in `match_rule_supported_vector`
* after changing note with @nsjian and @XiaohongGong , I think that two checks in `inline_vector_conver` is useless now. However, these checks impact other cpus, so I need more reviewers.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8265244](https://bugs.openjdk.java.net/browse/JDK-8265244): Remove useless comparation in LibraryCallKit::inline_vector_convert()

### Contributors
 * Wang Huang `<whuang@openjdk.org>`
 * Ai Jiaming `<aijiaming1@huawei.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3507/head:pull/3507` \
`$ git checkout pull/3507`

Update a local copy of the PR: \
`$ git checkout pull/3507` \
`$ git pull https://git.openjdk.java.net/jdk pull/3507/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3507`

View PR using the GUI difftool: \
`$ git pr show -t 3507`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3507.diff">https://git.openjdk.java.net/jdk/pull/3507.diff</a>

</details>

### CODE REVIEW

**Comment by Wanghuang-Huawei** [comment]
@iwanowww Can you do me a favor to review this patch? Thank you.

**Review by neliasso — COMMENTED** [review]
Please change the bug and PR title to something that describes the problem or solution.

**Comment by Wanghuang-Huawei** [comment]
> Please change the bug and PR title to something that describes the problem or solution.

Thank you for  your review. I will change that.

File: src/hotspot/cpu/aarch64/aarch64.ad [hunk 1]
```diff
@@ -2401,6 +2401,11 @@ const bool Matcher::match_rule_supported_vector(int opcode, int vlen, BasicType
   } else { // NEON
     // Special cases
     switch (opcode) {
+    case Op_VectorReinterpret:
+      if (bit_size != 64 && bit_size != 128) {
```
Comment (XiaohongGong) [inline src/hotspot/cpu/aarch64/aarch64.ad:2405]: bit_size < 64 ?

File: test/hotspot/jtreg/compiler/vectorapi/TestCast4STo2I.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+/*
+ * @test
+ * @bug 8265244
+ * @modules jdk.incubator.vector
+ * @run main/othervm compiler.vectorapi.TestCast4STo2I
+ */
+
+public class TestCast4STo2I {
+    static final VectorSpecies<Short> SPECIESs = ShortVector.SPECIES_64;
```
Comment (XiaohongGong) [inline test/hotspot/jtreg/compiler/vectorapi/TestCast4STo2I.java:39]: SPECIES_S?

File: test/hotspot/jtreg/compiler/vectorapi/TestCast4STo2I.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    static final VectorSpecies<Integer> SPECIESi = IntVector.SPECIES_64;
```
Comment (XiaohongGong) [inline test/hotspot/jtreg/compiler/vectorapi/TestCast4STo2I.java:40]: ditto

File: test/hotspot/jtreg/compiler/vectorapi/TestCast4STo2I.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        iv.intoArray(output, 0);
+    }
+
+    public static void main(String[] args) {
+        for (int i = 0; i < INVOC_COUNT; i++) {
+            testVectorCastS2I(as, ai);
+        }
+        for (int i = 0; i < SIZE; i++) {
+            System.out.print(ai[i] + ", ");
+        }
```
Comment (XiaohongGong) [inline test/hotspot/jtreg/compiler/vectorapi/TestCast4STo2I.java:72]: It's better to add the correctness testing as well to make sure the results are right.

File: src/hotspot/cpu/aarch64/aarch64.ad (hunk 1 above)
Comment (Wanghuang-Huawei) [reply src/hotspot/cpu/aarch64/aarch64.ad:2405]: Reinterpret's size should bit_size == 64 ||  bit_size == 128

**Review by theRealAph — COMMENTED** [review]
Please change the title of this to

8265244: AArch64: Remove useless comparation in LibraryCallKit::inline_vector_convert()

since it affects the AArch64 port. This makes the logs and the mailing lists much easier to read.

Thanks.

**Comment by Wanghuang-Huawei** [comment]
> Please change the title of this to
> 
> 8265244: AArch64: Remove useless comparation in LibraryCallKit::inline_vector_convert()
> 
> since it affects the AArch64 port. This makes the logs and the mailing lists much easier to read.
> 
> Thanks.

Thank you for your suggestion. There is only one question here:

 the comparation 
```
int num_elem_for_cast = MIN2(num_elem_from, Matcher::max_vector_size(elem_bt_to));
```
which will be removed here can effect all cpus.

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 5]
```diff
@@ -1450,7 +1450,7 @@ bool LibraryCallKit::inline_vector_convert() {
       // Since input and output number of elements are not consistent, we need to make sure we
       // properly size. Thus, first make a cast that retains the number of elements from source.
       // In case the size exceeds the arch size, we do the minimum.
```
Comment (theRealAph) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:1452]: If you're going to remove the MIN2 you must also change this comment.

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 6]
```diff
@@ -1469,7 +1469,7 @@ bool LibraryCallKit::inline_vector_convert() {
     } else if (num_elem_from > num_elem_to) {
       // Since number elements from input is larger than output, simply reduce size of input (we are supposed to
       // drop top elements anyway).
```
Comment (theRealAph) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:1471]: This one too?

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 5 above)
Comment (Wanghuang-Huawei) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:1452]: Sure. I will remove that in next commit.

**Pull request closed without merging**